Installation issue: hypre needs to be repackaged with autoconf@2.70 or newer to use oneapi fortran

<!-- Thanks for taking the time to report this build failure. To proceed with the report please:

1. Title the issue "Installation issue: <name-of-the-package>".
2. Provide the information required below.

We encourage you to try, as much as possible, to reduce your problem to the minimal example that still reproduces the issue. That would help us a lot in fixing it quickly and effectively! -->

### Steps to reproduce the issue

<!-- Fill in the exact spec you are trying to build and the relevant part of the error message -->

```console
$ spack spec hypre
Input spec
--------------------------------
hypre

Concretized
--------------------------------
hypre@2.20.0%oneapi@2021.2.0~complex~cuda~debug~int64~internal-superlu~mixedint+mpi~openmp+shared~superlu-dist cuda_arch=none patches=6e3336b1d62155f6350dfe42b0f9ea25d4fa0af60c7e540959139deb93a26059 arch=linux-opensuse_leap15-cascadelake
    ^mpich@3.3.2%oneapi@2021.2.0~argobots+fortran+hwloc+hydra+libxml2+pci+romio~slurm~verbs~wrapperrpath device=ch4 netmod=ofi patches=eb982de3366d48cbc55eb5e0df43373a45d9f51df208abf0835a72dc6c0b4774 pmi=pmi arch=linux-opensuse_leap15-cascadelake
    ^openblas@0.3.10%oneapi@2021.2.0~consistent_fpcsr~ilp64+locking+pic+shared patches=865703b4f405543bbd583413fdeff2226dfda908be33639276c06e5aa7ae2cae threads=none arch=linux-opensuse_leap15-cascadelake
$ spack install hypre

```

### Information on your system

<!-- Please include the output of `spack debug report` -->
* **Spack:** 0.16.1-1521-b56d65fce5
* **Python:** 3.6.12
* **Platform:** linux-opensuse_leap15-zen2
* **Concretizer:** original

<!-- If you have any relevant configuration detail (custom `packages.yaml` or `modules.yaml`, etc.) you can add that here as well. -->

Using this environment:
[spack.yaml.txt](https://github.com/spack/spack/files/6229932/spack.yaml.txt)

And this compiler stanza:

```
- compiler:
    spec: oneapi@2021.2.0
    paths:
      cc: /soft/<full path hidden per NDA>/oneapi/bin/icx
      cxx: /soft/<full path hidden per NDA>/oneapi/bin/icpx
      f77: /soft/<full path hidden per NDA>/oneapi/bin/ifx
      fc: /soft/<full path hidden per NDA>/oneapi/bin/ifx
    flags: {}
    operating_system: opensuse_leap15
    target: x86_64
    modules:
    - /soft/<full path hidden per NDA>/aurora_mpich/icc-sockets-debug/39.2
    environment: 
      append_path:
        LD_LIBRARY_PATH: /soft/<full path hidden per NDA>/compiler/20210130/linux/compiler/lib/intel64_lin
    extra_rpaths: []
```

Here is the offending piece from the configure log. Autoconf 2.69 causes the erroneous -loopopt=0 option to be generated, 2.70 does not:

```
configure:6510: /soft/<full path hidden per NDA>/mpich-ofi-sockets-icc-debug-gen9-drop39/bin/mpicc -o conftest -g -O2   conftest.c   -loopopt=0 -L/soft/restricted/CNDA/sdk/2020.12.15.1-ats/oneapi/compiler/20210130/linux/compiler/lib/intel64_lin -L/soft/<full path hidden per NDA>/oneapi/compiler/20210130/linux/lib -L/soft/packaging/spack-builds/linux-rhel7-x86_64/gcc-9.3.0/gcc-9.3.0-qfmcwfbuvnpn47zxjzfjvodzjl6reerh/lib/gcc/x86_64-pc-linux-gnu/9.3.0/../../../../lib64 -L/soft/packaging/spack-builds/linux-rhel7-x86_64/gcc-9.3.0/gcc-9.3.0-qfmcwfbuvnpn47zxjzfjvodzjl6reerh/lib/gcc/x86_64-pc-linux-gnu/9.3.0/../../../../lib64/ -L/lib/../lib64 -L/lib/../lib64/ -L/usr/lib/../lib64 -L/usr/lib/../lib64/ -L/soft/packaging/spack-builds/linux-rhel7-x86_64/gcc-9.3.0/binutils-2.34-rnwhrdgiqluqiypg5palnxdxviv3mynt/lib/../x86_64-pc-linux-gnu/lib/ -L/soft/packaging/spack-builds/linux-rhel7-x86_64/gcc-9.3.0/gcc-9.3.0-qfmcwfbuvnpn47zxjzfjvodzjl6reerh/lib/gcc/x86_64-pc-linux-gnu/9.3.0/../../../ -L/lib64 -L/lib/ -L/usr/lib64 -L/usr/lib -lifport -lifcoremt -limf -lsvml -lm -lipgo -lirc -lpthread -lirc_s -ldl >&5
/soft/packaging/spack-builds/linux-rhel7-x86_64/gcc-9.3.0/binutils-2.34-rnwhrdgiqluqiypg5palnxdxviv3mynt/bin/ld: cannot find -loopopt=0
clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
```

### Additional information

The issue is described on an Intel forum, and updating to autoconf 2.70 is the recommended fix:

https://community.intel.com/t5/Intel-Fortran-Compiler/ifx-2021-1-beta04-HPC-Toolkit-build-error-with-loopopt/m-p/1235375#M153121

<!-- Please upload the following files. They should be present in the stage directory of the failing build. Also upload any config.log or similar file if one exists. -->
* [spack-build-out.txt]()
* [spack-build-env.txt]()
These files contain NDA info and cannot be shared, however the above excerpt captures the issue.

<!-- Some packages have maintainers who have volunteered to debug build failures. Run `spack maintainers <name-of-the-package>` and @mention them here if they exist. -->
maintainers: @balay @osborn9 @ulrikeyang

### General information

<!-- These boxes can be checked by replacing [ ] with [x] or by clicking them after submitting the issue. -->
- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [*] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

Hello @frankwillmore same problem occurs for hdf5 :+1: 
spack install hdf5+fortran~mpi %oneapi
 * error in config.log : /usr/bin/ld: cannot find -loopopt=0
   * configure: error: linking to Fortran libraries from C fails

--- a/lib/spack/env/cc
+++ b/lib/spack/env/cc
@@ -311,6 +311,14 @@ while [ -n "$1" ]; do
             fi
             ;;
         -l*)
+            # -loopopt=0 is passed to the linker erroneously in
+            # autoconf <= 2.69. Filter it out.
+            # TODO: generalize filtering of args with an env var, so that
+            # TODO: we do not have to special case this here.
+            if [ "$mode" = "ld" ] && [ "$1" != "${1#-loopopt}" ]; then
+                shift
+                continue
+            fi
             arg="${1#-l}"
             if [ -z "$arg" ]; then shift; arg="$1"; fi
             other_args+=("-l$arg")

--- a/lib/spack/spack/test/cc.py
+++ b/lib/spack/spack/test/cc.py
@@ -622,3 +622,17 @@ def test_filter_enable_new_dtags(wrapper_flags):
         result = cc(*(test_args + ['-Wl,--enable-new-dtags']), output=str)
         result = result.strip().split('\n')
         assert '-Wl,--enable-new-dtags' not in result
+
+
+@pytest.mark.regression('22643')
+def test_linker_strips_loopopt(wrapper_flags):
+    with set_env(SPACK_TEST_COMMAND='dump-args'):
+        # ensure that -loopopt=0 is not present in ld mode
+        result = ld(*(test_args + ["-loopopt=0"]), output=str)
+        result = result.strip().split('\n')
+        assert '-loopopt=0' not in result
+
+        # ensure that -loopopt=0 is present in compile mode
+        result = cc(*(test_args + ["-loopopt=0"]), output=str)
+        result = result.strip().split('\n')
+        assert '-loopopt=0' in result

